### PR TITLE
Integrated weights and biases

### PR DESCRIPTION
**Pull request by sayannath**
Added `ModelCheckpoint`, `CSVLogger` and `WandbCallback`

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -1,17 +1,67 @@
...
+        filepath=f"{FLAGS.experiment_configs.checkpoint_filepath}_{timestamp}",
+        save_best_only=True,
+        save_weights_only=True,
+    )
+
+    logger_path = os.path.join(
+        "training_logs",
+        f"{FLAGS.experiment_configs.project_name}_{timestamp}",
+    )
+    logger_callback = keras.callbacks.CSVLogger(filepath=logger_path)
```
Comment (soumik12345) [inline train.py]: Any particular reason for using this?

File: train.py [hunk 2]
```diff
@@ -1,17 +1,67 @@
...
+    logging.info(f"Number of training examples: {pipeline_train.cardinality()}")
+    logging.info(f"Number of validation examples: {pipeline_test.cardinality()}")
+
+    # Setting up callback
+    logging.info("Initializing callbacks.")
+
+    # Model Checkpoint Callback
+    checkpoint_callback = keras.callbacks.ModelCheckpoint(
+        filepath=f"{FLAGS.experiment_configs.checkpoint_filepath}_{timestamp}",
+        save_best_only=True,
```
Comment (soumik12345) [inline train.py]: This is probably not required for a GAN model. Instead, the best course of action would be to create a custom callback to version each checkpoint as an [artifact](https://docs.wandb.ai/guides/artifacts). Unfortunately `wandb.keras.WandbCallback()` doesn't automatically save checkpoints yet, but it's gonna be resolved soon. Till then I think that we could create a custom callback inheriting from `keras.callbacks.ModelCheckpoint`. You can refer to [this](https://github.com/soumik12345/tf2_gans/blob/master/gaugan/callbacks.py#L74) implementation.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath I think we should also create a separate custom callback for visualizing the results during the training as a [wandb.Table](https://docs.wandb.ai/guides/data-vis/tables-quickstart). This would allow us to effortlessly showcase the transition in the quality of the outputs as the model gradually learns. You can refer to [this](https://github.com/soumik12345/tf2_gans/blob/master/gaugan/callbacks.py#L12) implementation.

File: train.py (hunk 1 above)
Comment (sayannath) [reply train.py]: For logging the loss results. Should I remove this?

Comment (soumik12345) [reply train.py]: Yeah, if you're using WandB, the `CSVLogger` seems redundant.

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath, There's only one thing left to be done. We should create an additional callback left to be created that would log a set of generated images to WandB. Otherwise, the PR looks good.

**Note:** Ayush and myself had created a PR for the WandB Keras integration to save models automatically as artifacts in savedmodel format. The PR has been merged to the client, but will be available for users from the next release. So, once that release is made, we can remove the additional artifacts callback.

**Comment by sayannath** [comment]
@soumik12345 added `GANMonitor` as a callback, but did not removed the `CheckpointArtifactCallback`. Will remove it in future.

File: sagan/callbacks.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        use_wandb: bool,
+        plot_save_dir: str,
+        num_images_plot: int = 4,
+    ):
+        self.val_images = next(iter(val_dataset))
+        self.n_samples = n_samples
+        self.epoch_interval = epoch_interval
+        self.plot_save_dir = plot_save_dir
+        self.use_wandb = use_wandb
+        self.columns = ["Semantic Mask", "Ground Truth", "Generated Image"]
```
Comment (soumik12345) [inline sagan/callbacks.py]: What are these???

File: sagan/callbacks.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        wandb_table = wandb.Table(columns=self.columns)
+        for i in range(self.num_images_plot):
+            wandb_table.add_data(
+                wandb.Image(semantic_input[i]),
+                wandb.Image(ground_truth[i]),
+                wandb.Image(generated_images[i]),
+            )
+        wandb.log({f"GANMonitor Epoch {epoch}": wandb_table})
+
+    def on_epoch_end(self, epoch, logs=None):
```
Comment (soumik12345) [inline sagan/callbacks.py:47]: I meant a GANMonitor callback for SAGAN, not the one for GauGAN 🙂

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath I think I didn't make the requirement for the GANMonitor callback clear enough. This callback is supposed to monitor the performance of the GAN and log the predictions of a single input batch over the epochs. Now, in our case, the GAN is not a conditional GAN like GauGAN, so there's no need to use tables. Simply go for something simple like `wandb.log({"Predictions": wandb.Image(pred_image)}, step=epoch)` inside `on_epoch_end`.

File: sagan/callbacks.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        use_wandb: bool,
+        plot_save_dir: str,
+        num_images_plot: int = 4,
+    ):
+        self.val_images = next(iter(val_dataset))
+        self.n_samples = n_samples
+        self.epoch_interval = epoch_interval
+        self.plot_save_dir = plot_save_dir
+        self.use_wandb = use_wandb
+        self.columns = ["Generated Image"]
```
Comment (soumik12345) [inline sagan/callbacks.py:27]: @sayannath A table with a single column has no use as a table. You can log the images directly without logging them as a table. Just use 

```python
for image in images:
        wandb.log({"Generated Images": wandb.Image(image_array, caption="Top: Output, Bottom: Input")}, step=epoch)
```

**Review by soumik12345 — CHANGES_REQUESTED** [review]
@sayannath please make the requested changes. otherwise the callback looks good.
